src/Backend/SgateShopgatePlugin/Controllers/Frontend/shopgate.php

```diff
@@ -1006,6 +1281,654 @@ public function registrationAction()
         }
     }
 
+    /**
+     * Custom action to get the address data of an user
+     */
+    public function getAddressesAction()
+    {
+        $decoded = $this->getJWT($this->Request()->getCookie('token'));
+        $customer = $this->getCustomer($decoded['customer_id']);
+
+        $defaultBillingAddress = $customer->getDefaultBillingAddress();
+        $defaultShippingAddress = $customer->getDefaultShippingAddress();
+
+        $addressRepository = Shopware()->Models()->getRepository("Shopware\\Models\\Customer\\Address");
+
+        $addresses = $addressRepository->getListArray($customer->getId());
+
+        // Create a list of ids of occurring countries and states
+        $countryIds = array_unique(array_filter(array_column($addresses, 'countryId')));
+        $stateIds = array_unique(array_filter(array_column($addresses, 'stateId')));
+
+        $countryRepository = $this->container->get('shopware_storefront.country_gateway');
+        $context = $this->container->get('shopware_storefront.context_service')->getShopContext();
+
+        $countries = $countryRepository->getCountries($countryIds, $context);
+        $states = $countryRepository->getStates($stateIds, $context);
+
+        // Apply translations for countries and states to address array, converting them from structs to arrays in the process
+        foreach ($addresses as &$address) {
+            if (array_key_exists($address['countryId'], $countries)) {
+                $address['country'] = json_decode(json_encode($countries[$address['countryId']]), true);
+            }
+            if (array_key_exists($address['stateId'], $states)) {
+                $address['state'] = json_decode(json_encode($states[$address['stateId']]), true);
+            }
+
+            $customerAddress = $addressRepository->getOneByUser($address['id'], $customer->getId());
+            $address['additional'] = $customerAddress->getAdditional();
+            $address['defaultBillingAddress'] = $defaultBillingAddress->getId() === $address['id'];
+            $address['defaultShippingAddress'] = $defaultShippingAddress->getId() === $address['id'];
+        }
+        unset($address);
+
+        $this->Response()->setHeader('Content-Type', 'application/json');
+        $this->Response()->setBody(json_encode($addresses));
+
+        $this->Response()->sendResponse();
+        exit();
+    }
+
+    public function addAddressAction()
```

Comment:
move all address functionality into a class of its own to reduce the complexity of this class(file)